### PR TITLE
Update go version to 1.20.4

### LINKED ISSUE #2054: publicly reported security vulnerabilities
Hi team,

In our vulnerability scan report there are 3 HIGH and 2 CRITICAL vulnerabilities in kube-state-metrics/kube-state-metrics:2.7.0

Component | Vulnerability | Severity | Installed Version | Latest Version | Description
| ---------- | ------------ | -------- | ----------------- | ---- | ---- |
| golang-runtime      | [CVE-2023-24538](https://nvd.nist.gov/vuln/detail/CVE-2023-24538) | CRITICAL | 1.19.4 | 1.20.3 | https://github.com/advisories/GHSA-v4m2-x4rp-hv22
|  golang-runtime  | [CVE-2023-24536](https://nvd.nist.gov/vuln/detail/CVE-2023-24536)   | HIGH | 1.19.4 | 1.20.3 | https://github.com/advisories/GHSA-9f7g-gqwh-jpf5
|  golang-runtime  | [CVE-2023-24534](https://nvd.nist.gov/vuln/detail/CVE-2023-24534)   | HIGH | 1.19.4 | 1.20.3 | https://github.com/advisories/GHSA-8v5j-pwr7-w5f8
|  golang-runtime | [CVE-2023-24537](https://nvd.nist.gov/vuln/detail/CVE-2023-24537)   | HIGH | 1.19.4 | 1.20.3  | https://github.com/advisories/GHSA-fp86-2355-v99r
|  etcd | [CVE-2021-28235](https://nvd.nist.gov/vuln/detail/CVE-2021-28235)  | CRITICAL | MISSING | v3.6.0-alpha.0  | https://github.com/advisories/GHSA-gmph-wf7j-9gcm

My suggestion: Golang version update to 1.20.3 should solve the problem.

### PR DESCRIPTION
**Pull request by RamakrishnanArun**
**What this PR does / why we need it**:
This PR upgrades the Go version to 1.20.3 and go-restful to 3.10.2 which resolves a number of security vulnerabilities.

**How does this change affect the cardinality of KSM**: *(increases, decreases or does not change cardinality)* Does not change

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2054

### CODE REVIEW

**Comment by RamakrishnanArun** [comment]
I've run the Twistlock tool against this and the report shows no vulnerabilities.

**Comment by mrueg** [comment]
I'm fine with the go 1.20.3 bump, I would like to see go-restful merged in https://github.com/kubernetes/kubernetes/pull/115067 (in particular because of https://github.com/kubernetes/kubernetes/pull/115067#issuecomment-1399119354 ) before we include this here.

File: go.sum [hunk 1]
```diff
@@ -64,6 +64,8 @@ github.com/dgryski/go-jump v0.0.0-20211018200510-ba001c3ffce0/go.mod h1:4hKCXuwr
 github.com/docopt/docopt-go v0.0.0-20180111231733-ee0de3bc6815/go.mod h1:WwZ+bS3ebgob9U8Nd0kOddGdZWjyMGR8Wziv+TBNwSE=
 github.com/emicklei/go-restful/v3 v3.9.0 h1:XwGDlfxEnQZzuopoqxwSEllNcCOM9DhhFyhFIIGKwxE=
 github.com/emicklei/go-restful/v3 v3.9.0/go.mod h1:6n3XBCmQQb25CM2LCACGz8ukIrRry+4bhvbpWn3mrbc=
+github.com/emicklei/go-restful/v3 v3.10.2 h1:hIovbnmBTLjHXkqEBUz3HGpXZdM7ZrE9fJIZIqlJLqE=
+github.com/emicklei/go-restful/v3 v3.10.2/go.mod h1:6n3XBCmQQb25CM2LCACGz8ukIrRry+4bhvbpWn3mrbc=
```
Comment (liam-verta) [inline go.sum:68]:
```suggestion
```

**Comment by liam-verta** [comment]
@RamakrishnanArun
Perhaps you could update the PR to remove (per the suggestions) to only update Go version? 

It would be better to get this merged and a new release out.

**Comment by RamakrishnanArun** [comment]
@liam-verta Thanks for your input. I think I misunderstood @mrueg 's comment. I understood it as a required change but only merged after the linked change in https://github.com/kubernetes/kubernetes/pull/115067 merged first.

So the reason for adding the change in both was that I got security vulnerabilities in both the go version and the version of go-restful that I am indirectly using with this project. That's why I wanted to add both updates.

Should I split them into 2 different PRs? One for the Go version and then one for the Go-Restful version?

File: Makefile [hunk 2]
```diff
@@ -16,7 +16,7 @@ OS ?= $(shell uname -s | tr A-Z a-z)
 ALL_ARCH = amd64 arm arm64 ppc64le s390x
 PKG = github.com/prometheus/common
 PROMETHEUS_VERSION = 2.42.0
-GO_VERSION = 1.20.2
+GO_VERSION = 1.20.3
```
Comment (mrueg) [inline Makefile:19]: go 1.20.4 got released yesterday, please use this version
https://groups.google.com/g/golang-announce/c/MEb0UyuSMsU

**Review by mrueg — COMMENTED** [review]
For go-restful, I would suggest to reach out on https://github.com/kubernetes/kubernetes/pull/115067 
I doubt this is actually exploitable here though.
We use https://go.dev/security/vuln/ govulncheck here, which checks the affected codepaths, most of the other scanners use a noisy approach of simply checking go.mod files, which is unreliable.

File: go.mod [hunk 3]
```diff
@@ -1,6 +1,6 @@
 module k8s.io/kube-state-metrics/v2
 
-go 1.19
+go 1.20
```
Comment (mrueg) [inline go.mod:3]: What's the intention in changing this here?

File: go.mod [hunk 4]
```diff
@@ -32,7 +32,7 @@ require (
 	github.com/cespare/xxhash/v2 v2.2.0 // indirect
 	github.com/coreos/go-systemd/v22 v22.5.0 // indirect
 	github.com/davecgh/go-spew v1.1.1 // indirect
-	github.com/emicklei/go-restful/v3 v3.9.0 // indirect
+	github.com/emicklei/go-restful/v3 v3.10.2 // indirect
```
[the remark this answers is not in the document]
Comment (RamakrishnanArun) [reply go.mod:35]: Rolling this back.

File: Makefile (hunk 2 above)
Comment (RamakrishnanArun) [reply Makefile:19]: Updated.

Comment (mrueg) [reply Makefile:19]: It seems like it is still at 1.20.3, can you push the change?

**Comment by RamakrishnanArun** [comment]
Thank you @liam-verta @mrueg for your review and inputs!

**Comment by mrueg** [comment]
One last request, since this change has become a single line now. Can you rebase everything into a single commit?

**Comment by RamakrishnanArun** [comment]
@mrueg done! Single commit now.

**Comment by mrueg** [comment]
/lgtm

Thanks!

**Comment by RamakrishnanArun** [comment]
When do you think the next release would be done?

**Comment by liam-verta** [comment]
@mrueg 
When do you think the next release would be done? This addresses reported CVE in dependencies so it would be good to cut a new image.

**Pull request merged**